src/compiler/manipulate_wasm_binary.ts

```diff
@@ -0,0 +1,244 @@
+
+// TODO put the licenses in the binary? Or with Azle? Probably with Azle actually
+// TODO it would be neat to be the licenses in all Azle binaries though
+// TODO can we make the start function just load the passive segment into memory?
+
+export async function manipulateWasmBinary(
+    canisterName: string,
+    js: string,
+    compilerInfo: CompilerInfo,
+    canisterConfig: CanisterConfig
+): Promise<void> {
+    const originalWasm = await readFile(
+        join(AZLE_PACKAGE_PATH, `static_canister_template.wasm`)
+    );
+
+    const module = binaryen.readBinary(originalWasm);
+
+    compilerInfo.canister_methods.queries.forEach(
+        ({ name: functionName, index, composite }) => {
+            addCanisterMethod(
+                module,
+                `${
+                    composite === true
+                        ? 'canister_composite_query'
+                        : 'canister_query'
+                } ${functionName}`,
+                'execute_js',
+                index,
+                true
+            );
+        }
+    );
+
+    compilerInfo.canister_methods.updates.forEach(
+        ({ name: functionName, index }) => {
+            addCanisterMethod(
+                module,
+                `canister_update ${functionName}`,
+                'execute_js',
+                index,
+                true
+            );
+        }
+    );
+
+    addCanisterMethod(
+        module,
+        'canister_init',
+        'init',
+        compilerInfo.canister_methods.init?.index ?? -1,
+        true
+    );
+
+    addCanisterMethod(
+        module,
+        'canister_post_upgrade',
+        'post_upgrade',
+        compilerInfo.canister_methods.post_upgrade?.index ?? -1,
+        true
+    );
+
+    if (compilerInfo.canister_methods.pre_upgrade !== undefined) {
+        addCanisterMethod(
+            module,
+            'canister_pre_upgrade',
+            'execute_js',
+            compilerInfo.canister_methods.pre_upgrade.index,
+            false
+        );
+    }
+
+    if (compilerInfo.canister_methods.inspect_message !== undefined) {
+        addCanisterMethod(
+            module,
+            'canister_inspect_message',
+            'execute_js',
+            compilerInfo.canister_methods.inspect_message.index,
+            true
+        );
+    }
+
+    if (compilerInfo.canister_methods.heartbeat !== undefined) {
+        addCanisterMethod(
+            module,
+            'canister_heartbeat',
+            'execute_js',
+            compilerInfo.canister_methods.heartbeat.index,
+            false
+        );
+    }
+
+    const memoryInfo = module.getMemoryInfo();
+
+    const numMemorySegments = module.getNumMemorySegments();
+
+    let segments = [];
+
+    for (let i = 0; i < numMemorySegments; i++) {
+        const segment = module.getMemorySegmentInfoByIndex(i);
+
+        segments.push(segment);
+    }
+
+    const normalizedSegments = segments.map((segment) => {
```

Comment:
why do we need to normalize and how does this do that?

Reply:
The data structures are just slightly different for some reason, so this just maps to the correct data structure. It's doing it right here, I read out of one data structure directly into the other.